openfisca_uk_data/datasets/frs/frs.py

```diff
@@ -1,87 +1,635 @@
+        .groupby(oddjob.person_id)
+        .sum()
+        .reindex(person.index)
+        .fillna(0)
+        + person.ALLPAY2
+        + person.ROYYR2
+        + person.ROYYR3
+        + person.ROYYR4
+        + person.CHAMTERN
+        + person.CHAMTTST
+    ) * 52
+
+    frs["private_transfer_income"] = (
+        person[
+            ["APAMT", "APDAMT", "PAREAMT", "ALLPAY1", "ALLPAY3", "ALLPAY4"]
+        ].sum(axis=1)
+    ) * 52
+
+    frs["lump_sum_income"] = person.REDAMT
+
+
+def add_benefit_income(
+    frs: h5py.File,
+    person: DataFrame,
+    benefits: DataFrame,
+    household: DataFrame,
+):
+    """Adds benefit variables.
+
+    Args:
+        frs (h5py.File)
+        person (DataFrame)
+        benefits (DataFrame)
+        household (DataFrame)
+    """
+    BENEFIT_CODES = dict(
+        child_benefit=3,
+        income_support=19,
+        housing_benefit=94,
+        AA=12,
+        DLA_SC=1,
+        DLA_M=2,
+        IIDB=15,
+        carers_allowance=13,
+        SDA=10,
+        AFCS=8,
+        maternity_allowance=21,
+        pension_credit=4,
+        child_tax_credit=91,
+        working_tax_credit=90,
+        state_pension=5,
+        winter_fuel_allowance=62,
+        incapacity_benefit=17,
+        universal_credit=95,
+        PIP_M=97,
+        PIP_DL=96,
+    )
+
+    for benefit, code in BENEFIT_CODES.items():
+        frs[benefit + "_reported"] = (
+            benefits.BENAMT[benefits.BENEFIT == code]
+            .groupby(benefits.person_id)
+            .sum()
+            .reindex(index=person.index)
+            .fillna(0)
+            .values
+            * 52
+        )
+
+    frs["BSP_reported"] = (
+        benefits.BENAMT[benefits.BENEFIT.isin((6, 9))]
+        .groupby(benefits.person_id)
+        .sum()
+        .reindex(index=person.index)
+        .fillna(0)
+        .values
+        * 52
+    )
+
+    frs["winter_fuel_allowance_reported"][...] = (
+        np.array(frs["winter_fuel_allowance_reported"]) / 52
+    )
+
+    frs["SSP_reported"] = person.SSPADJ * 52
+
+    frs["student_loans"] = person.TUBORR
+
+    frs["student_payments"] = person[["ADEMAAMT", "CHEMAAMT", "ACCSSAMT"]].sum(
+        axis=1
+    ) * 52 + person[["GRTDIR1", "GRTDIR2"]].sum(axis=1)
+
+    frs["council_tax_benefit_reported"] = (
+        (person.HRPID == 1)
+        * pd.Series(
+            household.CTREBAMT[person.household_id].values, index=person.index
+        ).fillna(0)
+        * 52
+    )
+
+
+def add_expenses(
+    frs: h5py.File,
+    person: DataFrame,
+    household: DataFrame,
+    maintenance: DataFrame,
+    mortgage: DataFrame,
+    childcare: DataFrame,
+    pen_prov: DataFrame,
+):
+    """Adds expense variables
+
+    Args:
+        frs (h5py.File)
+        person (DataFrame)
+        household (DataFrame)
+        maintenance (DataFrame)
+        mortgage (DataFrame)
+        childcare (DataFrame)
+        pen_prov (DataFrame)
+    """
+    frs["maintenance_expenses"] = (
+        pd.Series(
```

Comment:
i think you could save some code by making this pattern a function, e.g.
```
def sum_to_level(vals, dataset, level):
    reindex = person.index if level == "person" else household.index
    return pd.Series(vals).groupby(dataset[level + "_id"].sum().reindex(reindex).fillna(0)
```
i.e. this then becomes
```
frs["maintenance_expenses"] = sum_to_level(
    np.where(maintenance.MRUS == 2, maintenance.MRUAMT, maintenance.MRAMT),
    maintenance,
    "person") * 52